src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,469 @@
+    event Unwind(bytes32 indexed ilk, uint256 amount);
+    event Reap(bytes32 indexed ilk, uint256 amt);
+    event Collect(bytes32 indexed ilk, address indexed king, address[] assets, uint256 amt);
+    event Cage();
+    event Cage(bytes32 indexed ilk);
+    event Cull(bytes32 indexed ilk);
+    event Uncull(bytes32 indexed ilk);
+
+    constructor(address chainlog_) public {
+        address vat_ = ChainlogLike(chainlog_).getAddress("MCD_VAT");
+        address daiJoin_ = ChainlogLike(chainlog_).getAddress("MCD_JOIN_DAI");
+
+        chainlog = ChainlogLike(chainlog_);
+        vat = VatLike(vat_);
+        daiJoin = DaiJoinLike(daiJoin_);
+        dai = TokenLike(DaiJoinLike(daiJoin_).dai());
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "DssDirectDepositHub/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositHub/overflow");
+    }
+    uint256 constant RAY  = 10 ** 27;
+    function _rmul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, y) / RAY;
+    }
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Administration ---
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        D3M storage d3m = d3ms[ilk];
+        if (what == "bar") {
+            require(data <= d3m.join.getMaxBar(), "DssDirectDepositHub/above-max-interest");
+
+            d3m.bar = data;
+        } else if (what == "tau" ) {
+            require(live == 1, "DssDirectDepositHub/hub-not-live");
+            require(d3m.tic == 0, "DssDirectDepositHub/join-not-live");
+
+            d3m.tau = data;
+        } else revert("DssDirectDepositHub/file-unrecognized-param");
+
+        emit File(ilk, what, data);
+    }
+
+    function file(bytes32 ilk, bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
+        require(d3ms[ilk].tic == 0, "DssDirectDepositHub/join-not-live");
+
+        if (what == "king") d3ms[ilk].king = data;
+        else if (what == "join") d3ms[ilk].join = DssDirectDepositJoinLike(data);
+        else if (what == "gem") d3ms[ilk].gem = TokenLike(data);
+        else revert("DssDirectDepositHub/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+
+    // --- Deposit controls ---
+    function _wind(bytes32 ilk, DssDirectDepositJoinLike join, uint256 amount) internal {
```

Comment:
No need to provide the join since this contracts stores it and can access it. Or is it in order to spare a `SREAD` ?

```suggestion
    function _wind(bytes32 ilk, uint256 amount) internal {
```

Reply:
yes that is what i was thinking since i would need to look it up and store it again

Reply:
The gas savings are modest since the storage slot is [hot](https://www.evm.codes/#54) but on the other hand this is only an internal function so its interface is less relevant than what I first thought.

Reply:
ahh interesting.  i don't feel to strongly about it. do you? happy going either way since it doesn't seem to matter for gas or public interfaces

Reply:
I think you can leave it like that and save 100 gas